src/orchestration/orchestration.cpp

```diff
@@ -0,0 +1,911 @@
+        {
+            if (G.value->has_node(E.key))
+            {
+                orphaned = false;
+                break;
+            }
+        }
+
+        if (!orphaned)
+            continue;
+
+        WARN_PRINT(vformat("Removed orphan node %d from script %s.", E.key, get_self()->get_path()));
+        _nodes.erase(E.key);
+    }
+
+    for (const OScriptConnection& E : _connections)
+    {
+        if (!_nodes.has(E.from_node) || !_nodes.has(E.to_node))
+        {
+            WARN_PRINT("Orphan connection found: " + E.to_string());
+            //_connections.erase(E);
+        }
+    }
+}
+
+StringName Orchestration::get_base_type() const
+{
+    return _base_type;
+}
+
+void Orchestration::set_base_type(const StringName& p_base_type)
+{
+    if (!_base_type.match(p_base_type))
+    {
+        _base_type = p_base_type;
+        _self->emit_changed();
+    }
+}
+
+int Orchestration::get_available_id() const
+{
+    // We should eventually consider a better strategy for node unique ids to deal with
+    // scripts that are constantly modified with new nodes added and removed.
+
+    int max = -1;
+    for (const KeyValue<int, Ref<OScriptNode>>& E : _nodes)
+        max = Math::max(E.key, max);
+
+    return (max + 1);
+}
+
+void Orchestration::set_edited(bool p_edited)
+{
+    if (_edited != p_edited)
+    {
+        _edited = p_edited;
+        if (_edited)
+            _self->emit_changed();
+    }
+}
+
+void Orchestration::post_initialize()
+{
+    // Initialize variables
+    for (const KeyValue<StringName, Ref<OScriptVariable>>& E : _variables)
+        E.value->post_initialize();
+
+    // Initialize nodes
+    for (const KeyValue<int, Ref<OScriptNode>>& E : _nodes)
+        E.value->post_initialize();
+
+    // Initialize graphs
+    for (const KeyValue<StringName, Ref<OScriptGraph>>& G : _graphs)
+        G.value->post_initialize();
+
+    _initialized = true;
+}
+
+void Orchestration::validate_and_build(BuildLog& p_log)
+{
+    // Sanity check
+    _fix_orphans();
+
+    for (const KeyValue<int, Ref<OScriptNode>>& E : _nodes)
+    {
+        p_log.set_current_node(E.value);
+
+        E.value->validate_node_during_build(p_log);
+        p_log.set_current_node({});
+    }
+}
+
+bool Orchestration::has_node(int p_node_id) const
+{
+    return _nodes.has(p_node_id);
+}
+
+void Orchestration::add_node(const Ref<OScriptGraph>& p_graph, const Ref<OScriptNode>& p_node)
+{
+    ERR_FAIL_COND_MSG(_has_instances(), "Cannot add node, instances exist.");
+    ERR_FAIL_COND(p_node.is_null());
+    ERR_FAIL_COND(_nodes.has(p_node->get_id()));
+
+    // Validate the node details
+    p_node->_orchestration = this;
+    p_node->_validate_input_default_values();
+
+    // Register the node with the script
+    _nodes[p_node->get_id()] = p_node;
+
+    // Register the node with the graph
+    p_graph->add_node(p_node);
+}
+
+void Orchestration::remove_node(int p_node_id)
```

Comment:
Could `remove_node` only exist publically on `OScriptGraph` and then it has friend access to `Orchestration`?

Reply:
Going to leave this as-is.